apps/aeoracle/src/aeo_state_tree.erl

```diff
@@ -67,103 +80,131 @@ prune(Height, #oracle_tree{} = Tree) ->
 
-enter_common(Id, Expires, Serialized, Tree) ->
-    MTree1 = aeu_mtrees:enter(Id, Serialized, Tree#oracle_tree.mtree),
-    Cache = cache_push(Id, Expires, Tree#oracle_tree.cache),
-    Tree#oracle_tree{ mtree = MTree1
-                    , cache = Cache
+add_interaction(How, I, Tree) ->
+    OracleId    = aeo_interaction:oracle_address(I),
+    Oracle      = get_oracle(OracleId, Tree),
+    ITree       = gb_trees:get(OracleId, Tree#oracle_tree.itrees),
+    Id          = aeo_interaction:id(I),
+    SerializedI = aeo_interaction:serialize(I),
+    Expires     = aeo_interaction:expires(I),
+    ITree1      = case How of
+                      enter  -> aeu_mtrees:enter(Id, SerializedI, ITree);
+                      insert -> aeu_mtrees:insert(Id, SerializedI, ITree)
+                  end,
+
+    IRoot       = safe_root_hash(ITree1),
+    SerializedO = aeo_oracles:serialize(
+                      aeo_oracles:set_interactions_hash(IRoot, Oracle)),
+
+    OTree  = aeu_mtrees:enter(OracleId, SerializedO, Tree#oracle_tree.otree),
+    ITrees = gb_trees:update(OracleId, ITree1, Tree#oracle_tree.itrees),
+    Cache  = cache_push({interaction, OracleId, Id}, Expires, Tree#oracle_tree.cache),
+    Tree#oracle_tree{ otree  = OTree
+                    , itrees = ITrees
+                    , cache  = Cache
                     }.
 
-insert_common(Id, Expires, Serialized, Tree) ->
-    MTree1 = aeu_mtrees:insert(Id, Serialized, Tree#oracle_tree.mtree),
-    Cache = cache_push(Id, Expires, Tree#oracle_tree.cache),
-    Tree#oracle_tree{ mtree = MTree1
-                    , cache = Cache
-                    }.
+int_prune(Height, #oracle_tree{ cache = Cache } = Tree) ->
+    int_prune(cache_safe_peek(Cache), Height, Tree).
 
-int_prune(Height, #oracle_tree{cache = Cache, mtree = MTree} = Tree) ->
-    case cache_safe_peek(Cache) of
-        {H, _} when H > Height -> Tree;
-        Other ->
-            {Cache1, Mtree1} = int_prune(Other, Height, Cache, MTree),
-            Tree#oracle_tree{ cache = Cache1
-                            , mtree = Mtree1}
-    end.
-
-int_prune(none,_Height, Cache, MTree) ->
-    {Cache, MTree};
-int_prune({Height, Id}, Height, Cache, MTree) ->
+int_prune(none, _Height, Tree) ->
+    Tree;
+int_prune({Height, Id}, Height, #oracle_tree{ cache = Cache } = Tree) ->
     {{Height, Id}, Cache1} = cache_pop(Cache),
-    MTree1 = aeu_mtrees:delete(Id, MTree),
-    int_prune(cache_safe_peek(Cache1), Height, Cache1, MTree1);
-int_prune({Height1,_Id}, Height2, Cache, MTree) when Height2 < Height1 ->
-    {Cache, MTree}.
-
+    Tree1 = delete(Id, Tree#oracle_tree{ cache = Cache1 }),
+    int_prune(cache_safe_peek(Cache1), Height, Tree1);
+int_prune({Height1,_Id}, Height2, Tree) when Height2 < Height1 ->
+    Tree.
+
+delete({oracle, Id}, Tree) ->
+    OTree  = aeu_mtrees:delete(Id, Tree#oracle_tree.otree),
+    ITrees = gb_trees:delete(Id, Tree#oracle_tree.itrees),
```

Comment:
Shall any refunding take place? See https://github.com/aeternity/protocol/blob/f578d480f8c8e13c1425f26ebb732db01d099ab3/oracles/oracle_state_tree.md#pruning-of-oracle-interaction-objects

Reply:
See below.